### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by v1r3n**

### CODE REVIEW

File: pkg/workflow/def/workflow/do_while.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
+package workflow
+
+import (
+	"fmt"
+
+	"github.com/conductor-sdk/conductor-go/pkg/http_model"
+)
+
+var (
+	LOOP_CONDITION = "loop_count"
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/do_while.go:10]: nit: we should use MixedCaps or mixedCaps for the naming:  https://go.dev/doc/effective_go#Getters

File: pkg/workflow/def/workflow/do_while.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+)
+
+type DoWhileTask struct {
+	Task
+	loopCondition string
+	loopOver      []TaskInterface
+}
+
+func DoWhile(taskRefName string, terminationCondition string, tasks ...TaskInterface) *DoWhileTask {
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/do_while.go:19]: nit: `NewDoWhile(....)`

File: pkg/workflow/def/workflow/do_while.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+			inputParameters:   map[string]interface{}{},
+		},
+		loopCondition: terminationCondition,
+		loopOver:      tasks,
+	}
+}
+
+// Loop over N times when N is specified as iterations
+// can be  static number e.g. 5 or a parameter expression like ${task_ref.output.some_value} that is a number
+func Loop(taskRefName string, iterations int32, tasks ...TaskInterface) *DoWhileTask {
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/do_while.go:36]: nit: `NewLoop(...)`

File: pkg/workflow/def/workflow/fork_join.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package workflow
+
+import "github.com/conductor-sdk/conductor-go/pkg/http_model"
+
+type ForkTask struct {
+	Task
+	forkedTasks [][]TaskInterface
+}
+
+func Fork(taskRefName string, tasks ...[]TaskInterface) *ForkTask {
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/fork_join.go:10]: nit: `NewFork(...)`

File: pkg/workflow/def/workflow/fork_join.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+			taskType:          FORK_JOIN,
+			optional:          false,
+			inputParameters:   map[string]interface{}{},
+		},
+		forkedTasks: tasks,
+	}
+}
+
+func (task *ForkTask) toWorkflowTask() []http_model.WorkflowTask {
+	forkWorkflowTask := task.Task.toWorkflowTask()[0]
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/fork_join.go:25]: should we check the length before accessing to the first element?

File: pkg/workflow/def/workflow/fork_join_dynamic.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import "github.com/conductor-sdk/conductor-go/pkg/http_model"
+
+type DynamicForkTask struct {
+	Task
+	preForkTask TaskInterface
+	join        JoinTask
+}
+
+func DynamicFork(taskRefName string, forkPrepareTask TaskInterface) *DynamicForkTask {
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/fork_join_dynamic.go:11]: nit: `NewDynamicFork(...`

File: pkg/workflow/def/workflow/fork_join_dynamic.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+			description:       "",
+			taskType:          FORK_JOIN_DYNAMIC,
+			optional:          false,
+			inputParameters:   map[string]interface{}{},
+		},
+		preForkTask: forkPrepareTask,
+	}
+}
+
+func DynamicForkWithJoin(taskRefName string, forkPrepareTask TaskInterface, join JoinTask) *DynamicForkTask {
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/fork_join_dynamic.go:25]: nit: New...

File: pkg/workflow/def/workflow/fork_join_dynamic.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+			inputParameters:   map[string]interface{}{},
+		},
+		preForkTask: forkPrepareTask,
+		join:        join,
+	}
+}
+
+func (task *DynamicForkTask) toWorkflowTask() []http_model.WorkflowTask {
+
+	forkWorkflowTask := task.Task.toWorkflowTask()[0]
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/fork_join_dynamic.go:42]: check for length?

File: pkg/workflow/def/workflow/fork_join_dynamic.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+	forkWorkflowTask.DynamicForkTasksParam = "forkedTasks"
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/fork_join_dynamic.go:43]: nit: make `forkedTasks` a variable on top of the file

File: pkg/workflow/def/workflow/http.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+	GET     HttpMethod = "GET"
+	PUT     HttpMethod = "PUT"
+	POST    HttpMethod = "POST"
+	DELETE  HttpMethod = "DELETE"
+	HEAD    HttpMethod = "HEAD"
+	OPTIONS HttpMethod = "OPTIONS"
+)
+
+func Http(taskRefName string, input *HttpInput) *HttpTask {
+	if input.Method == "" {
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/http.go:19]: nit: `if len(input.Method) == 0 {`

File: pkg/workflow/def/workflow/http.go [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+const (
+	GET     HttpMethod = "GET"
+	PUT     HttpMethod = "PUT"
+	POST    HttpMethod = "POST"
+	DELETE  HttpMethod = "DELETE"
+	HEAD    HttpMethod = "HEAD"
+	OPTIONS HttpMethod = "OPTIONS"
+)
+
+func Http(taskRefName string, input *HttpInput) *HttpTask {
```
Comment (mdarabi) [inline pkg/workflow/def/workflow/http.go:18]: nit: New...

**Review by mdarabi — COMMENTED** [review]
Overall it looks good but there are multiple nits that should be addressed (following the effective go instructions: https://go.dev/doc/effective_go)

- Constructors should start with word `New`
- Ensure all the constant values are captured at the top of the file using `var` or `const`
- Follow the naming conventions as mentioned here: https://go.dev/doc/effective_go
- There are some ` // TODO` that must be addressed
- There are places where we have a nested map e.g. `map[string]map[string]...` these are really hard to understand. we usually convert a `map[string]` to a struct so the final type would look like as `map[string]struct1`
- To avoid panics, we should check the length of the slice unless you are 100% sure the underlying slice cannot be ever modified`

**Comment by mdarabi** [comment]
Also in general we need to handle panics in the SDK, so we need to have a recover method in the top level

**Comment by v1r3n** [comment]
> Also in general we need to handle panics in the SDK, so we need to have a recover method in the top level

Do you have a suggestion or example on this?

File: pkg/worker/task_runner.go [hunk 12]
```diff
@@ -17,75 +19,123 @@ import (
...
 	)
+	return NewTaskRunnerWithApiClient(apiClient)
+}
+
+func NewTaskRunnerWithApiClient(
+	apiClient *conductor_http_client.APIClient,
+) *TaskRunner {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = ""
```
Comment (mdarabi) [inline pkg/worker/task_runner.go:43]: not sure if it would cause problems on the server side or not. so in other words, should we error out here?

**Comment by mdarabi** [comment]
> > Also in general we need to handle panics in the SDK, so we need to have a recover method in the top level
> 
> Do you have a suggestion or example on this?

One method that can benefit from the Panic Recovery is the task runner ( which i think is one of the core part of the sdk) where there is a high chance of panic (concurrency, task execution, etc). Here is just a sample example: https://gobyexample.com/recover

In my experience wherever I am not sure what will be happening, i add the recover method (specifically api calls and multi-threading) to ensure my code would not cause the service goes down and handles the panic properly.

File: pkg/worker/task_runner.go (hunk 12 above)
Comment (v1r3n) [reply pkg/worker/task_runner.go:45]: This is just for the informational purpose, so if we cannot get hostname, its okay.

File: pkg/workflow/def/workflow/do_while.go (hunk 1 above)
Comment (v1r3n) [reply pkg/workflow/def/workflow/do_while.go]: fixed.

File: pkg/workflow/def/workflow/fork_join.go (hunk 5 above)
Comment (v1r3n) [reply pkg/workflow/def/workflow/fork_join.go:25]: toWorkflowTask in the Task always returns an array with 1 element, so no need.

File: pkg/workflow/def/workflow/fork_join_dynamic.go (hunk 8 above)
Comment (v1r3n) [reply pkg/workflow/def/workflow/fork_join_dynamic.go:46]: toWorkflowTask in the Task always returns an array with 1 element, so no need.

File: pkg/workflow/def/workflow/fork_join_dynamic.go (hunk 9 above)
Comment (v1r3n) [reply pkg/workflow/def/workflow/fork_join_dynamic.go]: fixed.

File: pkg/workflow/def/workflow/http.go (hunk 10 above)
Comment (v1r3n) [reply pkg/workflow/def/workflow/http.go]: fixed.

**Review by gardusig — APPROVED** [review]

**Pull request merged**